File: data/figures/chapter02/figure_2_8/plot_2_8.m
function plot_2_8(outputFile)
%PLOT_2_8 Reproduce thesis Figure 2.8 from the sibling CSV only.
scriptFile = mfilename('fullpath');
figureDirectory = fileparts(scriptFile);
dataFile = fullfile(figureDirectory, 'figure_2_8.csv');
if nargin < 1 || strlength(string(outputFile)) == 0
    outputFile = fullfile(figureDirectory, 'plot_2_8.png');
else
    outputFile = char(string(outputFile));
end
outputDirectory = fileparts(outputFile);
if ~isempty(outputDirectory) && ~isfolder(outputDirectory)
    mkdir(outputDirectory);
end

data = readtable(dataFile, 'TextType', 'string', ...
    'VariableNamingRule', 'preserve');
required = ["configuration", "root", "velocity_mps", ...
    "damping_g", "positive"];
assert(all(ismember(required, string(data.Properties.VariableNames))), ...
    'Figure 2.8 CSV has an unexpected schema.');
assert(height(data) == 1480, 'Expected 1,480 valid V-g samples.');
assert(isequal(unique(data.root, 'sorted').', 1:20), ...
    'Expected retained roots 1 through 20.');

configurationNames = ["coarse_edge_length_2p5_mm", ...
    "geometry_tolerance_2p0_mm"];
configurationTitles = ["Coarse edge length: 2.5 mm", ...
    "Geometry tolerance: 2.0 mm"];
panelLetters = ["a", "b"; "c", "d"];

figureHandle = figure('Visible', 'off', 'Color', 'white', ...
    'Units', 'inches', 'Position', [0.5, 0.35, 7.25, 10.75], ...
    'Renderer', 'painters');
layout = tiledlayout(figureHandle, 2, 2, ...
    'TileSpacing', 'loose', 'Padding', 'loose');
title(layout, 'Case 64 V-g diagnostic', 'FontName', 'Times New Roman', ...
    'FontSize', 13, 'FontWeight', 'bold', 'Interpreter', 'none');

for rowIndex = 1:2
    rows = data(data.configuration == configurationNames(rowIndex), :);
    assert(height(rows) == 740, ...
        'Expected 740 samples for each Figure 2.8 configuration.');
    plotAllRoots(nexttile(layout), rows, ...
        panelLetters(rowIndex, 1) + '. ' + ...
        configurationTitles(rowIndex) + ', all roots');
    plotRoot19(nexttile(layout), rows, ...
        panelLetters(rowIndex, 2) + '. ' + ...
        configurationTitles(rowIndex) + ', root 19');
end

exportgraphics(figureHandle, outputFile, 'Resolution', 220, ...
    'BackgroundColor', 'white');
close(figureHandle);
fprintf('%s\n', outputFile);
end

function plotAllRoots(ax, data, panelTitle)
roots = unique(data.root, 'sorted').';
colors = lines(numel(roots));
hold(ax, 'on');
for index = 1:numel(roots)
    rows = data(data.root == roots(index), :);
    rows = sortrows(rows, 'velocity_mps', 'ascend');
    plot(ax, rows.velocity_mps, rows.damping_g, '-', ...
        'Color', colors(index, :), 'LineWidth', 0.85, ...
        'HandleVisibility', 'off');
end
drawZeroLine(ax);
hold(ax, 'off');
formatAxes(ax, panelTitle, false);
end

function plotRoot19(ax, data, panelTitle)
rows = data(data.root == 19, :);
rows = sortrows(rows, 'velocity_mps', 'ascend');
positive = logical(rows.positive);
hold(ax, 'on');
plot(ax, rows.velocity_mps, rows.damping_g, '-', ...
    'Color', [0.00, 0.31, 0.55], 'LineWidth', 1.15, ...
    'DisplayName', 'Root 19');
scatter(ax, rows.velocity_mps(positive), rows.damping_g(positive), 28, ...
    'o', 'MarkerFaceColor', [0.80, 0.16, 0.12], ...
    'MarkerEdgeColor', [0.80, 0.16, 0.12], ...
    'DisplayName', 'Positive sample');
drawZeroLine(ax);
hold(ax, 'off');
formatAxes(ax, panelTitle, true);
legend(ax, 'Location', 'southwest', 'FontName', 'Times New Roman', ...
    'FontSize', 8.5, 'Interpreter', 'none', 'Box', 'off');
end

function drawZeroLine(ax)
yline(ax, 0, '-', 'Color', [0.10, 0.10, 0.10], 'LineWidth', 1.1, ...
    'HandleVisibility', 'off');
end

function formatAxes(ax, panelTitle, scientificTicks)
xlim(ax, [30, 150]);
grid(ax, 'on');
box(ax, 'on');
title(ax, panelTitle, 'FontName', 'Times New Roman', 'FontSize', 10, ...
    'FontWeight', 'normal', 'Interpreter', 'none');
xlabel(ax, 'Velocity, V [m/s]', 'FontName', 'Times New Roman', ...
    'FontSize', 9.5, 'Interpreter', 'none');
ylabel(ax, 'Damping, g [-]', 'FontName', 'Times New Roman', ...
    'FontSize', 9.5, 'Interpreter', 'none');
set(ax, 'FontName', 'Times New Roman', 'FontSize', 9, ...
    'TickLabelInterpreter', 'none', 'TickDir', 'out', 'LineWidth', 0.8, ...
    'Layer', 'top', 'XGrid', 'on', 'YGrid', 'on', ...
    'GridLineStyle', ':', 'GridAlpha', 0.18);
ax.YAxis.Exponent = 0;
if scientificTicks
    ytickformat(ax, '%.1e');
else
    ytickformat(ax, '%.2f');
end
end

File: data/figures/chapter02/figure_2_8/plot_2_8.py
"""Reproduce thesis Figure 2.8 from the sibling CSV only."""

import argparse
import csv
from pathlib import Path

import matplotlib.pyplot as plt
from matplotlib.ticker import FormatStrFormatter


HERE = Path(__file__).resolve().parent
DATA_FILE = HERE / "figure_2_8.csv"
DEFAULT_OUTPUT_FILE = HERE / "plot_2_8.png"

CONFIGURATIONS = [
    ("coarse_edge_length_2p5_mm", "Coarse edge length: 2.5 mm"),
    ("geometry_tolerance_2p0_mm", "Geometry tolerance: 2.0 mm"),
]


def format_axis(axis, title, scientific=False):
    axis.set_xlim(30, 150)
    axis.axhline(0.0, color="0.1", linewidth=1.1)
    axis.set_title(title, loc="left", fontsize=10, fontweight="normal")
    axis.set_xlabel("Velocity, V [m/s]")
    axis.set_ylabel("Damping, g [-]")
    axis.grid(True, linestyle=":", alpha=0.25)
    if scientific:
        axis.set_ylim(-1.6e-5, 2.0e-6)
        axis.yaxis.set_major_formatter(FormatStrFormatter("%.1e"))
    else:
        axis.set_ylim(-0.16, 0.02)


def plot_all_roots(axis, data, title):
    roots = sorted({row["root"] for row in data})
    colors = plt.get_cmap("tab20").colors
    for index, root in enumerate(roots):
        rows = sorted(
            (row for row in data if row["root"] == root),
            key=lambda row: row["velocity_mps"],
        )
        axis.plot(
            [row["velocity_mps"] for row in rows],
            [row["damping_g"] for row in rows],
            linewidth=0.85,
            color=colors[index % len(colors)],
        )
    format_axis(axis, title)


def plot_root_19(axis, data, title):
    rows = sorted(
        (row for row in data if row["root"] == 19),
        key=lambda row: row["velocity_mps"],
    )
    axis.plot(
        [row["velocity_mps"] for row in rows],
        [row["damping_g"] for row in rows],
        linewidth=1.15,
        color=(0.00, 0.31, 0.55),
        label="Root 19",
    )
    positive = [row for row in rows if row["positive"]]
    axis.scatter(
        [row["velocity_mps"] for row in positive],
        [row["damping_g"] for row in positive],
        s=28,
        color=(0.80, 0.16, 0.12),
        label="Positive sample",
        zorder=3,
    )
    format_axis(axis, title, scientific=True)
    axis.legend(loc="lower left", frameon=False, fontsize=8)


def parse_args():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--output",
        type=Path,
        default=DEFAULT_OUTPUT_FILE,
        help="Output image file (default: sibling plot_2_8.png)",
    )
    return parser.parse_args()


def main(output_file=DEFAULT_OUTPUT_FILE):
    required = {"configuration", "root", "velocity_mps", "damping_g", "positive"}
    with DATA_FILE.open(newline="", encoding="utf-8") as stream:
        reader = csv.DictReader(stream)
        missing = required.difference(reader.fieldnames or [])
        if missing:
            raise ValueError(f"Missing required columns: {sorted(missing)}")
        data = []
        for source in reader:
            data.append(
                {
                    "configuration": source["configuration"],
                    "root": int(source["root"]),
                    "velocity_mps": float(source["velocity_mps"]),
                    "damping_g": float(source["damping_g"]),
                    "positive": source["positive"].strip().lower()
                    in {"1", "true"},
                }
            )
    if len(data) != 1480:
        raise ValueError(f"Expected 1,480 valid V-g samples, found {len(data)}")
    if sorted({row["root"] for row in data}) != list(range(1, 21)):
        raise ValueError("Expected retained roots 1 through 20")

    plt.rcParams.update({"font.family": "Times New Roman", "font.size": 9})
    figure, axes = plt.subplots(2, 2, figsize=(7.25, 10.75), constrained_layout=True)
    figure.suptitle("Case 64 V-g diagnostic", fontsize=13, fontweight="bold")

    for row_index, (configuration, title) in enumerate(CONFIGURATIONS):
        subset = [row for row in data if row["configuration"] == configuration]
        if len(subset) != 740:
            raise ValueError(
                f"Expected 740 samples for {configuration}, found {len(subset)}"
            )
        plot_all_roots(
            axes[row_index, 0],
            subset,
            f"{'a' if row_index == 0 else 'c'}. {title}, all roots",
        )
        plot_root_19(
            axes[row_index, 1],
            subset,
            f"{'b' if row_index == 0 else 'd'}. {title}, root 19",
        )

    output_file = Path(output_file).expanduser().resolve()
    output_file.parent.mkdir(parents=True, exist_ok=True)
    figure.savefig(output_file, dpi=220, facecolor="white")
    plt.close(figure)
    print(output_file)


if __name__ == "__main__":
    arguments = parse_args()
    main(arguments.output)
